### PR TITLE
Add saml login to backstage

### PR DESCRIPTION
**Pull request by jibone**
## Adds SAML login to Backstage

Hey, I'm trying to add SAML login to Backstage, as our organization's identity providers uses SAML. I saw that there is an issue created for SAML integration #1891 with tag "help wanted" if there is no SAML integration yet I would like to contribute. 

Here was my thought process on this, I try to use the common provider but found that it is more bias towards an OAuth2 flow. The more I work on this the more I think perhaps it should be a seperate provider? I'm looking more towards user login via SAML identity provider.

Screenshot 2020-09-04 at 7 02 27 PM

Screenshot 2020-09-22 at 2 17 54 PM

This current PR works on local with the local SAML script "start-saml-idp.sh". There's several places that I need to fix and clean up. However am I going in the right direction though? Would appreatiate any advice :)

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: packages/core-api/src/apis/definitions/auth.ts [hunk 1]
```diff
@@ -97,6 +97,24 @@ export type OAuthApi = {
...
+export type SamlApi = {
+  // Not sure what Promise call back should have.
+  getBackstageIdentity(
+    options?: AuthRequestOptions,
+  ): Promise<BackstageIdentity | undefined>;
+
+  getProfile(options?: AuthRequestOptions): Promise<ProfileInfo | undefined>;
+
+  // FIXME: is this needed?
+  getAccessToken(options?: AuthRequestOptions): Promise<string>;
```
Comment (soapraj) [inline packages/core-api/src/apis/definitions/auth.ts:112]: Can be removed when there is no access token flow in SAML Auth.

File: packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      id: payload.profile.email,
+    };
+  }
+
+  // FIXME: do we need this for SAML?
+  async refreshSession(): Promise<any> {}
+
+  async removeSession(): Promise<void> {
+    // FIXME: remove hardcoded url here... should do something like buildUrl
+    const res = await fetch('http://localhost:7000/auth/saml/logout', {
```
Comment (soapraj) [inline packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts:80]: +1

File: packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    });
+
+    return {
+      ...payload,
+      id: payload.profile.email,
+    };
+  }
+
+  // FIXME: do we need this for SAML?
+  async refreshSession(): Promise<any> {}
```
Comment (soapraj) [inline packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts:76]: Not required - can be removed.

File: packages/core-api/src/lib/AuthSessionManager/types.ts [hunk 4]
```diff
@@ -36,6 +36,13 @@ export type SessionManager<T> = {
   sessionState$(): Observable<SessionState>;
 };
 
+// FIXME do we need this?
+export type SamlAuthSessionManager<T> = {
```
Comment (soapraj) [inline packages/core-api/src/lib/AuthSessionManager/types.ts:40]: I could be wrong but the type is similar to SessionManager<T> and don't see a specific need for this.

File: packages/core/src/layout/Sidebar/UserSettings.tsx [hunk 5]
```diff
@@ -36,6 +36,12 @@ export function SidebarUserSettings({
     if (!sidebarOpen && open) setOpen(false);
   }, [open, sidebarOpen]);
 
+  // FIXME: Change this and remove the session storage stuff
+  const handleLogout = () => {
```
Comment (soapraj) [inline packages/core/src/layout/Sidebar/UserSettings.tsx:40]: Why do we have to clear the sessionStorage here?

File: plugins/auth-backend/src/providers/saml/provider.ts [hunk 6]
```diff
@@ -53,8 +53,10 @@ export class SamlAuthProvider implements AuthProviderRouteHandlers {
       //       for non-oauth auth flows.
       // TODO: This flow doesn't issue an identity token that can be used to validate
       //       the identity of the user in other backends, which we need in some form.
+      console.log('===> SamlAuthProvider constructor');
```
Comment (soapraj) [inline plugins/auth-backend/src/providers/saml/provider.ts:56]: remove console logs

**Comment by soapraj** [comment]
Great work! Thanks for the PR and welcome to the community.

You are certainly on the right track and i left comments around the FIXME's in the code. Let me know if there is anything else that I need to look into. 

For getting an entry in the sidebar for SAML you may want to look in [here](https://github.com/spotify/backstage/blob/master/packages/core/src/layout/Sidebar/DefaultProviderSettings.tsx) and create something similar to [OIDCProviderSettings](https://github.com/spotify/backstage/blob/f2bb9be15f5ede3a9d7fdcc31ee3b54752b513b5/packages/core/src/layout/Sidebar/Settings/OIDCProviderSettings.tsx) component. Then you can decide based on the app-config.yaml if you have saml config and render the sidebar item.

I spun it up locally and the saml auth flow works fine :)

**Comment by jibone** [comment]
Thank you @soapraj I was away for these few days. I'll work on FIXME's :)

File: packages/core-api/src/lib/AuthSessionManager/SamlAuthSessionManager.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  SamlAuthConnector,
+  SamlResponse,
+} from '../AuthConnector/SamlAuthConnector';
+import { SessionStateTracker } from './SessionStateTracker';
+
+type Options = {
+  connector: SamlAuthConnector<SamlResponse>;
+};
+
+export class SamlAuthSessionManager<T> implements SessionManager<T> {
```
Comment (Rugvip) [inline packages/core-api/src/lib/AuthSessionManager/SamlAuthSessionManager.ts:28]: Is this the same as the original? These helper classes are intended to be generic and use-case oriented. Only copy if you need them to support something new, and in the case name it accordingly

File: packages/core-api/src/lib/AuthSessionManager/SamlAuthSessionStore.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+import { SamlAuthSessionManager, GetSessionOptions } from './types';
+
+type Options<T> = {
+  manager: SamlAuthSessionManager<T>;
+  storageKey: string;
+};
+
+export class SamlAuthSessionStore<T> implements SamlAuthSessionManager<T> {
```
Comment (Rugvip) [inline packages/core-api/src/lib/AuthSessionManager/SamlAuthSessionStore.ts:24]: Same thing here, is this addition needed? How is it different from the existing session store?

Comment (jibone) [reply packages/core-api/src/lib/AuthSessionManager/SamlAuthSessionStore.ts:24]: Yeah, I thought at first that it is going to be different, but then I think can reuse the generic session manager and session store. Will work on that.

Thanks.

**Comment by jibone** [comment]
@soapraj @Rugvip Thank you for reviewing this. I've made the recomended changes and this PR is ready for review.

**Comment by jibone** [comment]
Looks like there is a recent change on SessionApi. I'll fix this PR.

**Review by Rugvip — COMMENTED** [review]
Oh meh, yep, forgot to post these comments 😅

File: packages/core/src/layout/SignInPage/types.ts [hunk 9]
```diff
@@ -43,9 +44,9 @@ export type ProviderComponent = ComponentType<
 
 export type ProviderLoader = (
   apis: ApiHolder,
-  apiRef: ApiRef<
-    OAuthApi & ProfileInfoApi & BackstageIdentityApi & SessionStateApi
-  >,
+  apiRef:
+    | ApiRef<OAuthApi & ProfileInfoApi & BackstageIdentityApi & SessionStateApi>
+    | ApiRef<SamlApi>,
```
[the remark this answers is not in the document]
Comment (jibone) [reply packages/core/src/layout/SignInPage/types.ts]: Yup, just pulled the latest and saw `SessionApi`, I'm working towards that. Thanks 😄

**Comment by jibone** [comment]
@Rugvip Thanks, I've made the changes to use `sessionApi` and remove some stuff that is not needed. This PR is ready for review.

File: packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  provider: AuthProvider & { id: string };
+};
+
+export type SamlResponse = {
+  userId: string;
+  profile: ProfileInfo;
+  backstageIdentity: BackstageIdentity;
+};
+
+export class SamlAuthConnector<SamlResponse> {
```
Comment (Rugvip) [inline packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts:36]: Think we could rename this to `DirectAuthConnector` or so? since it can be used for many other strategies than just SAML. Also thinking we should highlight in the docs that this can only be used to implement direct users interaction flows like the sign-in page, since it lacks the extra dialog layer that ensures that the popup is opened through a user action.

File: packages/core-api/src/lib/AuthSessionManager/AuthSessionStore.ts [hunk 11]
```diff
@@ -28,7 +28,7 @@ type Options<T> = {
   /** Storage key to use to store sessions */
   storageKey: string;
   /** Used to get the scope of the session */
-  sessionScopes: SessionScopesFunc<T>;
+  sessionScopes?: SessionScopesFunc<T>;
```
Comment (Rugvip) [inline packages/core-api/src/lib/AuthSessionManager/AuthSessionStore.ts:31]: Yeah this ends up being a bit awkward without the dedicated SAML classes, still better though but could be something to refactor if we get a couple of other non-OAuth providers in place.

**Review by Rugvip — APPROVED** [review]
Thanks! good-to-go with a minor comment 👍

**Comment by benjdlambert** [comment]
@Rugvip could you take another look at this one too?

File: packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts (hunk 10 above)
Comment (jibone) [reply packages/core-api/src/lib/AuthConnector/SamlAuthConnector.ts]: 👌 made the changes to the class name

File: packages/core-api/src/lib/AuthSessionManager/AuthSessionStore.ts (hunk 11 above)
Comment (jibone) [reply packages/core-api/src/lib/AuthSessionManager/AuthSessionStore.ts:31]: Yup. Definitely something that can be refactored. Who knows, someone might implement LDAP in the future 😄

**Comment by jibone** [comment]
PR updated to address the comment 🙇

**Review by benjdlambert — APPROVED** [review]

**Comment by riginoommen** [comment]
I am trying to use this but docs seems to be insufficent

**Comment by suuus** [comment]
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻

**Comment by riginoommen** [comment]
How to use the SAML login any example
